# numpy notes 

## ch01 intro

In [1]:

intro="""
1. numpy is a python library. NumPy was created in 2005 by Travis Oliphant 
2. numpy stands for numerical python 
3. numpy is open source (free)
4. universal standard for working with numerical data in Python
5. The NumPy API is used extensively in Pandas, SciPy, Matplotlib, scikit-learn, scikit-image and most other data science 
   and scientific Python packages.
   
   Data Science: is a branch of computer science where we study how to store, 
   use and analyze data for deriving information from it.
   
"""


In [6]:
"""
The source code for NumPy is located at this github repository https://github.com/numpy/numpy
github: enables many people to work on the same codebase

If you already have Python, you can install NumPy with:

conda install numpy
or

pip install numpy

"""

'\nThe source code for NumPy is located at this github repository https://github.com/numpy/numpy\ngithub: enables many people to work on the same codebase\n\nIf you already have Python, you can install NumPy with:\n\nconda install numpy\nor\n\npip install numpy\n\n'

In [7]:
pip install numpy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.3.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
"""
6. NumPy aims to provide an array object that is up to 50x more faster than traditional Python lists.(see demo below)
"""

import numpy as np 

numbers=np.random.rand(10000000) #generate random numbers 

%timeit sum(numbers)  #python sum function 

%timeit np.sum(numbers) #numpy sum functions 

#defining a user defined function to add sum of random numbers and check calculation speed 

def mysum(num):
    total=0
    for i in num:
        total+=i
    return total
#now calling the function mysum 
%timeit mysum(numbers)

512 ms ± 8.82 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
8.24 ms ± 188 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
602 ms ± 2.81 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [3]:
"""
 NumPy, array data is stored in a continuous block of memory. This is a key feature that contributes to the efficiency 
 and speed of NumPy operations.

The ndarray object in NumPy represents a multidimensional, homogeneous array of fixed-size items. 
The elements in the array are stored in a contiguous block of memory, and the information about the shape, 
data type, and strides is stored in a separate data structure. 
The contiguous memory layout allows for efficient access and manipulation of array elements.
"""


import numpy as np

# Creating a NumPy array
arr = np.array([[21,22,25,67],
                [34,35,65,66],
                [22,24,67,45]])

# Checking the memory layout
#flags['C_CONTIGUOUS'] attribute checks whether the array is stored in C-style contiguous memory
print(arr.flags['C_CONTIGUOUS']) 

True


In [5]:
"""
The array object in NumPy is called ndarray
Arrays are very frequently used in data science, where speed and resources are very important.
"""
import numpy as np 
list1=[12,14,15,67,55]
print("the l1 data type is = ",type(list1))
arr=np.array(list1)
print("the arr data type is = ",type(arr))

the l1 data type is =  <class 'list'>
the arr data type is =  <class 'numpy.ndarray'>


# #ch02 basics

In [8]:
"""
in this chaper cover basic of scalar(single quantity),vector(1d array),matrix(2d array) and 
covers 3d array (matrix of matrixes), 
nd array, and conversion 1d to 2d array and vice versa 
ndarray.ndim, ndarray.size, ndarray.shape
np.array(), np.zeros(), np.ones(), np.empty(), np.arange(), np.linspace(), dtype (data types)

"""

'\nin this chaper cover basic of scalar(single quantity),vector(1d array),matrix(2d array) and \ncovers 3d array (matrix of matrixes), \nnd array, and conversion 1d to 2d array and vice versa \nndarray.ndim, ndarray.size, ndarray.shape\nnp.array(), np.zeros(), np.ones(), np.empty(), np.arange(), np.linspace(), dtype (data types)\n\n'

In [9]:
#scalar it has signle quantity has no dimensions (dimension =0)

import numpy as np 

a=20
arr=np.array(a)
#printing a value 
print(a)
print(type(a))
#printing arr value 
print(arr)
print(type(arr))
print(arr.ndim) #print the dimension of numpy array 
print(arr.shape) #print the shape of array (row,columns) in tuple format
print(arr.size) # print total number of elements present in array 

20
<class 'int'>
20
<class 'numpy.ndarray'>
0
()
1


In [10]:
#vector it has signle direction has one dimension (dimension =1) either row or columns vectors
#in numpy it is neither a row vector nor a column vector by default. It is just a one-dimensional array.
#its dimension is 1

import numpy as np 

a=[12,15,34,67,32,85]
arr=np.array(a)
#printing a value 
print(a)
print(type(a))
#printing arr value 
print(arr)
print(type(arr))
print(arr.ndim) #print the dimension of numpy array 
print(arr.shape) #print the shape of array (row,columns) in tuple format
print(arr.size) # print total number of elements present in array 

[12, 15, 34, 67, 32, 85]
<class 'list'>
[12 15 34 67 32 85]
<class 'numpy.ndarray'>
1
(6,)
6


In [11]:
#However, if you want to explicitly convert it to a row or column vector, you can use the reshape method. For example:
#now its dimension will be 2 
import numpy as np

a = [12, 15, 34, 67, 32, 85]
print("Original array:")
print(arr)

#convert to np array
arr = np.array(a)

# Convert to a row vector
row_vector = arr.reshape(1, -1)
print("\nRow vector:")
print(row_vector)
print(type(row_vector))
print(row_vector.ndim) #print the dimension of numpy array 
print(row_vector.shape) #print the shape of array (row,columns) in tuple format
print(row_vector.size) # print total number of elements present in array 

# Convert to a column vector
column_vector = arr.reshape(-1, 1)
print("\nColumn vector:")
print(column_vector)
print(type(column_vector))
print(column_vector.ndim) #print the dimension of numpy array 
print(column_vector.shape) #print the shape of array (row,columns) in tuple format
print(column_vector.size) # print total number of elements present in array 



Original array:
[12 15 34 67 32 85]

Row vector:
[[12 15 34 67 32 85]]
<class 'numpy.ndarray'>
2
(1, 6)
6

Column vector:
[[12]
 [15]
 [34]
 [67]
 [32]
 [85]]
<class 'numpy.ndarray'>
2
(6, 1)
6


In [12]:
#matrix it has two direction called row and columns vectors, it has dimension 2
#in numpy it is neither a row vector nor a column vector by default. It is just a one-dimensional array.
#its dimension is 2 it has both rows and colums more than 1 
#note if row =1 columns more than one than it will be colummn matrix (1X6) in previous example
#note if row =morethan one  columns =1 than it will be colummn matrix (6X1) in previous example 

#2d matrix 

import numpy as np

# Using np.array with a list of lists
matrix = np.array([
                    [1, 2, 3],
                    [4, 5, 6], 
                    [7, 8, 9]
                ])

print(matrix)


print("\n\nprinting another matrix\n\n ")
# Using np.array with a list of lists
matrix = np.array([[31, 23, 43], [94, 55, 86], [77, 88, 91],[34,65,78]])

print(matrix)
print("\nthe type of matix is =",type(matrix).__name__)
print("the matrix dimension is = ",matrix.ndim) #print the dimension of numpy array 
print("matrix shape in tuple = ",matrix.shape) #print the shape of array (row,columns) in tuple format
print("elements present in matrix is = ",matrix.size) # print total number of elements present in array 

print("printing matrix row and columns values ")
print("row = ",matrix.shape[0])
print("col = ",matrix.shape[1])


[[1 2 3]
 [4 5 6]
 [7 8 9]]


printing another matrix

 
[[31 23 43]
 [94 55 86]
 [77 88 91]
 [34 65 78]]

the type of matix is = ndarray
the matrix dimension is =  2
matrix shape in tuple =  (4, 3)
elements present in matrix is =  12
printing matrix row and columns values 
row =  4
col =  3


In [13]:
#creating 3 dimension using np.zeros()
import numpy as np

# Creating a 3-dimensional matrix filled with zeros
shape = (3, 3, 4)  # Specify the size of each dimension
matrix = np.zeros(shape)

print(matrix)
print("\nthe type of matix is =",type(matrix).__name__)
print("the matrix dimension is = ",matrix.ndim) #print the dimension of numpy array 
print("matrix shape in tuple = ",matrix.shape) #print the shape of array (row,columns) in tuple format
print("elements present in matrix is = ",matrix.size) # print total number of elements present in array 



[[[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]

 [[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]

 [[0. 0. 0. 0.]
  [0. 0. 0. 0.]
  [0. 0. 0. 0.]]]

the type of matix is = ndarray
the matrix dimension is =  3
matrix shape in tuple =  (3, 3, 4)
elements present in matrix is =  36


# matrix creation np.array(), np.zeros(), np.ones(), np.empty(), np.arange(), np.linspace()

In [14]:
#create 3X3 matrix with np.array()
import numpy as np 

matrix=np.array([[1,2,3],[4,5,6],[7,8,9]])
print(matrix)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [15]:
#create 3X3 matrix with np.zeros() it will give all the elements zero
#shapes should give in tuple format

import numpy as np 
#ex1
matrix=np.zeros((3,3))
print(matrix)
print("\n\n")
#ex2
shape=(4,3)
mat=np.zeros(shape)
print(mat)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]



[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


In [16]:
#create 3X3 matrix with np.ones() it will give all the elements one
#shapes should give in tuple format

import numpy as np 
#ex1
matrix=np.ones((3,3))
print(matrix)
print("\n\n")
#ex2
shape=(4,3)
mat=np.ones(shape)
print(mat)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]



[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]


In [17]:
#creating a 5 dimension array 
import numpy as np

arr = np.array([1, 2, 3, 4], ndmin=5)

print(arr)
print('number of dimensions :', arr.ndim)

[[[[[1 2 3 4]]]]]
number of dimensions : 5


In [18]:
# Creating an empty array with shape (2, 3)
#it will contains some garbage values 
import numpy as np

# Creating an empty array with shape (2, 3)
empty_array = np.empty((2, 3))

print(empty_array)

[[4.88059032e-313 1.99467604e-312 1.82491638e-312]
 [1.86735630e-312 7.21478569e-313 1.65515672e-312]]


In [19]:
#another way to create 
#order: Specifies whether to store the array in C-style (row-major) or Fortran-style (column-major) order. Optional, 
#and if not specified, it defaults to 'C'.
#
#
import numpy 
shape=(5,4)
temp=numpy.empty(shape, dtype=float, order='C')
print(temp)

[[0.00000000e+000 1.58101007e-322 0.00000000e+000 0.00000000e+000]
 [            nan 4.94065646e-324 9.48606040e-321 4.94065646e-324]
 [1.30438919e-311 1.30438919e-311 1.30436965e-311 0.00000000e+000]
 [0.00000000e+000 9.48606040e-321 0.00000000e+000 1.30437745e-311]
 [9.48606040e-321 1.30438914e-311 1.23516411e-322 2.12199579e-314]]


In [20]:
#create arrays with np.arange(start:end:step) 
#end ends with n-1 ex: in this code end =11, but it stops at 10
#step is default is 1. if you want we can change step 
import numpy as np

arange_array=np.arange(1,11)
print(arange_array)

arange_array=np.arange(11)
print(arange_array)

arange_array=np.arange(11,22)
print(arange_array)

arange_array=np.arange(11,22,2)
print(arange_array)

arange_array=np.arange(11,22,3)
print(arange_array)

[ 1  2  3  4  5  6  7  8  9 10]
[ 0  1  2  3  4  5  6  7  8  9 10]
[11 12 13 14 15 16 17 18 19 20 21]
[11 13 15 17 19 21]
[11 14 17 20]


In [21]:
#reshape the np.range() values 
arange_array=np.arange(1,11)
print(arange_array)


[ 1  2  3  4  5  6  7  8  9 10]


In [22]:
arange_array=np.arange(1,11).reshape((2,5))
print(arange_array)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]]


In [23]:
arange_array=np.arange(1,11).reshape((5,2))
print(arange_array)


[[ 1  2]
 [ 3  4]
 [ 5  6]
 [ 7  8]
 [ 9 10]]


##  linspace 

In [2]:
import numpy as np 

np.linspace(0,10,3)

array([ 0.,  5., 10.])

In [3]:
np.linspace(0,10,50)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

In [4]:
np.eye(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

## random 

In [5]:
np.random.rand(2)


array([0.31536052, 0.16538269])

In [6]:
np.random.randn(2)

array([-1.73018256,  0.65573219])

In [7]:
np.random.randn(5,5)

array([[ 0.40652316,  0.11817081,  0.91989206,  0.864321  ,  0.4458718 ],
       [ 0.61370651,  0.12166332,  1.5120274 , -1.39705121, -0.14582844],
       [ 0.43613662, -0.29615609,  1.0922124 , -1.04518803, -1.79317641],
       [ 1.04093237,  0.5640042 ,  0.60943018, -1.43280433,  0.47630341],
       [ 0.28105619,  1.02942581, -0.0205102 , -0.72540293, -1.09807097]])

In [8]:
np.random.randint(1,100)

43

In [9]:
np.random.randint(1,100,10)

array([80, 31, 84, 63, 21, 92, 37, 62, 98, 65])

##  Array Attributes and Methods

In [10]:
arr = np.arange(25)
ranarr = np.random.randint(0,50,10)

In [11]:
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [12]:
ranarr

array([36, 29,  4, 48,  6, 10, 29, 44,  7, 23])

In [13]:
arr.reshape(5,5)

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19],
       [20, 21, 22, 23, 24]])

In [14]:
ranarr

array([36, 29,  4, 48,  6, 10, 29, 44,  7, 23])

In [15]:
ranarr.max()

48

In [16]:
ranarr.argmax()

3

In [17]:
ranarr.min()

4

In [18]:
ranarr.argmin()

2

In [19]:
# Vector
arr.shape

(25,)

In [20]:
# Notice the two sets of brackets
arr.reshape(1,25)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
        16, 17, 18, 19, 20, 21, 22, 23, 24]])

In [21]:
arr.reshape(1,25).shape

(1, 25)

In [22]:
arr.reshape(25,1)

array([[ 0],
       [ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10],
       [11],
       [12],
       [13],
       [14],
       [15],
       [16],
       [17],
       [18],
       [19],
       [20],
       [21],
       [22],
       [23],
       [24]])

In [23]:
arr.reshape(25,1).shape

(25, 1)

In [24]:
arr.dtype

dtype('int32')

# NumPy Indexing and Selection

In [25]:
import numpy as np

In [26]:
#Creating sample array
arr = np.arange(0,11)

In [27]:
#Show
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

## Bracket Indexing and Selection

In [28]:
#Setting a value with index range (Broadcasting)
arr[0:5]=100

#Show
arr

array([100, 100, 100, 100, 100,   5,   6,   7,   8,   9,  10])

In [29]:
# Reset array, we'll see why I had to reset in  a moment
arr = np.arange(0,11)

#Show
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [30]:
#Important notes on Slices
slice_of_arr = arr[0:6]

#Show slice
slice_of_arr

array([0, 1, 2, 3, 4, 5])

In [31]:
#Change Slice
slice_of_arr[:]=99

#Show Slice again
slice_of_arr

array([99, 99, 99, 99, 99, 99])

In [32]:
arr

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

In [33]:
#Now note the changes also occur in our original array!
#Data is not copied, it's a view of the original array! This avoids memory problems!

#To get a copy, need to be explicit
arr_copy = arr.copy()

arr_copy

array([99, 99, 99, 99, 99, 99,  6,  7,  8,  9, 10])

# Indexing a 2D array (matrices)
The general format is arr_2d[row][col] or arr_2d[row,col]. I recommend usually using the comma notation for clarity.

In [34]:
arr_2d = np.array(([5,10,15],[20,25,30],[35,40,45]))

#Show
arr_2d

array([[ 5, 10, 15],
       [20, 25, 30],
       [35, 40, 45]])

In [35]:
#Indexing row
arr_2d[1]

array([20, 25, 30])

In [36]:
# Format is arr_2d[row][col] or arr_2d[row,col]

# Getting individual element value
arr_2d[1][0]

20

In [37]:
# Getting individual element value
arr_2d[1,0]

20

In [38]:
# 2D array slicing

#Shape (2,2) from top right corner
arr_2d[:2,1:]

array([[10, 15],
       [25, 30]])

In [39]:
#Shape bottom row
arr_2d[2]

array([35, 40, 45])

In [40]:
#Shape bottom row
arr_2d[2,:]

array([35, 40, 45])

# Fancy Indexing
Fancy indexing allows you to select entire rows or columns out of order,to show this, let's quickly build out a numpy array:

In [41]:
#Set up matrix
arr2d = np.zeros((10,10))

In [45]:
arr2d

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3., 3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4.],
       [5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],
       [6., 6., 6., 6., 6., 6., 6., 6., 6., 6.],
       [7., 7., 7., 7., 7., 7., 7., 7., 7., 7.],
       [8., 8., 8., 8., 8., 8., 8., 8., 8., 8.],
       [9., 9., 9., 9., 9., 9., 9., 9., 9., 9.]])

In [43]:
arr2d.shape

(10, 10)

In [42]:
#Length of array
arr_length = arr2d.shape[1]

In [44]:
#Set up array

for i in range(arr_length):
    arr2d[i] = i
    
arr2d

array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [3., 3., 3., 3., 3., 3., 3., 3., 3., 3.],
       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4.],
       [5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],
       [6., 6., 6., 6., 6., 6., 6., 6., 6., 6.],
       [7., 7., 7., 7., 7., 7., 7., 7., 7., 7.],
       [8., 8., 8., 8., 8., 8., 8., 8., 8., 8.],
       [9., 9., 9., 9., 9., 9., 9., 9., 9., 9.]])

In [46]:
arr2d[[2,4,6,8]]

array([[2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4.],
       [6., 6., 6., 6., 6., 6., 6., 6., 6., 6.],
       [8., 8., 8., 8., 8., 8., 8., 8., 8., 8.]])

In [47]:
#Allows in any order
arr2d[[6,4,2,7]]

array([[6., 6., 6., 6., 6., 6., 6., 6., 6., 6.],
       [4., 4., 4., 4., 4., 4., 4., 4., 4., 4.],
       [2., 2., 2., 2., 2., 2., 2., 2., 2., 2.],
       [7., 7., 7., 7., 7., 7., 7., 7., 7., 7.]])

# More Indexing Help
Indexing a 2d matrix can be a bit confusing at first, especially when you start to add in step size. Try google image searching NumPy indexing to fins useful images, like this one:

# Selection
Let's briefly go over how to use brackets for selection based off of comparison operators.

In [48]:
arr = np.arange(1,11)
arr

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10])

In [49]:
arr > 4

array([False, False, False, False,  True,  True,  True,  True,  True,
        True])

In [50]:
bool_arr = arr>4

In [51]:
bool_arr

array([False, False, False, False,  True,  True,  True,  True,  True,
        True])

In [52]:
arr[bool_arr]

array([ 5,  6,  7,  8,  9, 10])

In [53]:
arr[arr>2]

array([ 3,  4,  5,  6,  7,  8,  9, 10])

In [54]:
x = 2
arr[arr>x]

array([ 3,  4,  5,  6,  7,  8,  9, 10])

# NumPy Operations
Arithmetic
You can easily perform array with array arithmetic, or scalar with array arithmetic. Let's see some examples:

In [55]:
import numpy as np
arr = np.arange(0,10)

In [56]:
arr + arr

array([ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18])

In [57]:
arr * arr

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

In [58]:
arr - arr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [59]:
# Warning on division by zero, but not an error!
# Just replaced with nan
arr/arr

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\2878212635.py:3: RuntimeWarning: invalid value encountered in divide
  arr/arr


array([nan,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.,  1.])

In [60]:
# Also warning, but not an error instead infinity
1/arr

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\1360216608.py:2: RuntimeWarning: divide by zero encountered in divide
  1/arr


array([       inf, 1.        , 0.5       , 0.33333333, 0.25      ,
       0.2       , 0.16666667, 0.14285714, 0.125     , 0.11111111])

In [61]:
arr**3

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729], dtype=int32)

# Universal Array Functions
Numpy comes with many universal array functions, which are essentially just mathematical operations you can use to perform the operation across the array. Let's show some common ones:

In [62]:
#Taking Square Roots
np.sqrt(arr)

array([0.        , 1.        , 1.41421356, 1.73205081, 2.        ,
       2.23606798, 2.44948974, 2.64575131, 2.82842712, 3.        ])

In [63]:
#Calcualting exponential (e^)
np.exp(arr)

array([1.00000000e+00, 2.71828183e+00, 7.38905610e+00, 2.00855369e+01,
       5.45981500e+01, 1.48413159e+02, 4.03428793e+02, 1.09663316e+03,
       2.98095799e+03, 8.10308393e+03])

In [64]:
np.max(arr) #same as arr.max()

9

In [65]:
np.sin(arr)


array([ 0.        ,  0.84147098,  0.90929743,  0.14112001, -0.7568025 ,
       -0.95892427, -0.2794155 ,  0.6569866 ,  0.98935825,  0.41211849])

In [66]:
np.log(arr)

C:\Users\admin\AppData\Local\Temp\ipykernel_14876\3120950136.py:1: RuntimeWarning: divide by zero encountered in log
  np.log(arr)


array([      -inf, 0.        , 0.69314718, 1.09861229, 1.38629436,
       1.60943791, 1.79175947, 1.94591015, 2.07944154, 2.19722458])

# NumPy Exercises
Now that we've learned about NumPy let's test your knowledge. We'll start off with a few simple tasks, and then you'll be asked some more complicated questions.

Import NumPy as np

In [68]:
import numpy as np 

 Create an array of 10 zeros

In [69]:
arr=np.ones(10)
arr

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

Create an array of 10 ones

In [71]:
arr=np.ones(10)
arr

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

Create an array of 10 fives

In [72]:
arr=np.ones(10)*5
arr

array([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.])

Create an array of the integers from 10 to 50

Create an array of the integers from 10 to 50

In [75]:
arr=np.arange(10,51)
arr

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49, 50])

Create an array of all the even integers from 10 to 50

In [76]:
arr=np.arange(10,51,2)
arr

array([10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42,
       44, 46, 48, 50])

Create a 3x3 matrix with values ranging from 0 to 8

In [78]:
arr=np.arange(0,9).reshape(3,3)
arr

array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])

Create a 3x3 identity matrix

In [80]:
arr=np.eye(3)
arr

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

# random
Use NumPy to generate a random number between 0 and 1

In [81]:
import numpy as np 
from numpy.random import randn


In [82]:
arr =randn(1)
arr

array([-0.30329803])

Use NumPy to generate an array of 25 random numbers sampled from a standard normal distribution

In [83]:
a=np.random.randn(25)
a

array([-0.17947687,  0.99991378, -0.11230839,  0.82280102,  0.39167335,
       -0.04087085, -0.97742326, -0.41576347, -1.00792841, -1.50888936,
       -1.2262644 ,  0.21884957,  2.30462925,  0.62787038, -0.71281522,
       -0.37903332, -0.71412805,  0.4206816 , -1.00746884, -0.66148334,
       -1.09917404,  0.57653834, -0.16501008,  0.05461837,  0.92514791])

Create the following matrix:

array([[ 0.01,  0.02,  0.03,  0.04,  0.05,  0.06,  0.07,  0.08,  0.09,  0.1 ],
       [ 0.11,  0.12,  0.13,  0.14,  0.15,  0.16,  0.17,  0.18,  0.19,  0.2 ],
       [ 0.21,  0.22,  0.23,  0.24,  0.25,  0.26,  0.27,  0.28,  0.29,  0.3 ],
       [ 0.31,  0.32,  0.33,  0.34,  0.35,  0.36,  0.37,  0.38,  0.39,  0.4 ],
       [ 0.41,  0.42,  0.43,  0.44,  0.45,  0.46,  0.47,  0.48,  0.49,  0.5 ],
       [ 0.51,  0.52,  0.53,  0.54,  0.55,  0.56,  0.57,  0.58,  0.59,  0.6 ],
       [ 0.61,  0.62,  0.63,  0.64,  0.65,  0.66,  0.67,  0.68,  0.69,  0.7 ],
       [ 0.71,  0.72,  0.73,  0.74,  0.75,  0.76,  0.77,  0.78,  0.79,  0.8 ],
       [ 0.81,  0.82,  0.83,  0.84,  0.85,  0.86,  0.87,  0.88,  0.89,  0.9 ],
       [ 0.91,  0.92,  0.93,  0.94,  0.95,  0.96,  0.97,  0.98,  0.99,  1.  ]])

In [87]:
arr=np.arange(1,101).reshape(10,10)/100
arr

array([[0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 ],
       [0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 ],
       [0.21, 0.22, 0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 ],
       [0.31, 0.32, 0.33, 0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 ],
       [0.41, 0.42, 0.43, 0.44, 0.45, 0.46, 0.47, 0.48, 0.49, 0.5 ],
       [0.51, 0.52, 0.53, 0.54, 0.55, 0.56, 0.57, 0.58, 0.59, 0.6 ],
       [0.61, 0.62, 0.63, 0.64, 0.65, 0.66, 0.67, 0.68, 0.69, 0.7 ],
       [0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77, 0.78, 0.79, 0.8 ],
       [0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88, 0.89, 0.9 ],
       [0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99, 1.  ]])

Create an array of 20 linearly spaced points between 0 and 1:

In [88]:
arr=np.linspace(0,1,20)
arr

array([0.        , 0.05263158, 0.10526316, 0.15789474, 0.21052632,
       0.26315789, 0.31578947, 0.36842105, 0.42105263, 0.47368421,
       0.52631579, 0.57894737, 0.63157895, 0.68421053, 0.73684211,
       0.78947368, 0.84210526, 0.89473684, 0.94736842, 1.        ])

# Numpy Indexing and Selection
Now you will be given a few matrices, and be asked to replicate the resulting matrix outputs:

In [89]:
mat = np.arange(1,26).reshape(5,5)
mat

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [92]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[2:,1:]

array([[12, 13, 14, 15],
       [17, 18, 19, 20],
       [22, 23, 24, 25]])

In [96]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[3,4]

20

In [102]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[:3,1:2]

array([[ 2],
       [ 7],
       [12]])

In [105]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[-1]

array([21, 22, 23, 24, 25])

In [107]:
# WRITE CODE HERE THAT REPRODUCES THE OUTPUT OF THE CELL BELOW
# BE CAREFUL NOT TO RUN THE CELL BELOW, OTHERWISE YOU WON'T
# BE ABLE TO SEE THE OUTPUT ANY MORE
mat[3:,]

array([[16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

# Now do the following

Get the sum of all the values in mat

In [109]:
mat.sum()

325

325

Get the standard deviation of the values in mat

In [111]:
mat.std()

7.211102550927978

7.2111025509279782

Get the sum of all the columns in mat

In [116]:
mat

array([[ 1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10],
       [11, 12, 13, 14, 15],
       [16, 17, 18, 19, 20],
       [21, 22, 23, 24, 25]])

In [117]:
mat.sum(axis=0)

array([55, 60, 65, 70, 75])

array([55, 60, 65, 70, 75])